state/unit_test.go

```diff
@@ -614,6 +618,9 @@ func (s *UnitSuite) TestRemoveUnitMachineNoDestroyCharmProfile(c *gc.C) {
 
 	host, err := s.State.AddMachine("quantal", state.JobHostUnits)
 	c.Assert(err, jc.ErrorIsNil)
+	err = host.SetProvisioned("inst-id", "fake_nonce", nil)
+	c.Assert(err, jc.ErrorIsNil)
+
```

Comment:
@SimonRichardson, this test is failing at line 636.  I'm not sure if this part is still valid or not.  Thoughts?

Reply:
I'll have a look, from your PR.

Reply:
I'm wondering if we change the `unit.go` to the following (see below), that way we don't need to check for provisioning issues. Although this is a bit more defensive in terms of programming style, it does mean that we don't have to care as much about the `m.CharmProfiles()` machineProfile errors.

```
diff --git a/state/unit.go b/state/unit.go
index 1f0072e..620367e 100644
--- a/state/unit.go
+++ b/state/unit.go
@@ -716,18 +716,21 @@ func (u *Unit) destroyHostOps(a *Application) (ops []txn.Op, err error) {
                // if this is a unit removal from a machine which will not be
                // destroyed.
                machineProfiles, err := m.CharmProfiles()
-               if err != nil {
-                       return nil, err
-               }
-               profile, err := lxdprofile.MatchProfileNameByAppName(machineProfiles, u.ApplicationName())
-               if err != nil {
-                       return nil, err
-               }
-               if profile != "" {
-                       // TODO (hml) 09-jan-2019
-                       // how do we know when to clean this doc up?
-                       logger.Tracef("Setup to remove charm profile %q, removing unit from machine %s", profile, m.Id())
-                       ops = append(ops, m.SetUpgradeCharmProfileOp(a.Name(), "", lxdprofile.EmptyStatus))
+               if err == nil {
+                       profile, err := lxdprofile.MatchProfileNameByAppName(machineProfiles, u.ApplicationName())
+                       if err != nil {
+                               return nil, err
+                       }
+                       if profile != "" {
+                               // TODO (hml) 09-jan-2019
+                               // how do we know when to clean this doc up?
+                               logger.Tracef("Setup to remove charm profile %q, removing unit from machine %s", profile, m.Id())
+                               ops = append(ops, m.SetUpgradeCharmProfileOp(a.Name(), "", lxdprofile.EmptyStatus))
+                       }
+               } else {
+                       if !errors.IsNotProvisioned(err) {
+                               return nil, err
+                       }
                }
        }
```

Reply:
So the test now fails because we never call `SetUpgradeCharmProfileOp`. So the `chAppName` will always equal `""`. We should seriously think if this is what we intend.

Reply:
chAppName is always equal to "" here, because I need to set some charm profiles in the instance charm profile data.  :-)  overnight allowed for a light bulb moment to happen.

Reply:
To your code example, I had been thinking of a similar thing.  However that felt like coding for the unit tests to work, not because it was truly needed.  I'm up in the air on that.

Reply:
We'll keep this the way it currently is, as the tests shouldn't be short circuiting actual real world setup. We can always use this as a plan B if required.